### PR TITLE
Feature/allow benchmark and vintage selection

### LINKED ISSUE #3: Most TigerWeb data keys are unable to be populated
Due to the module _only_ sourcing its data from the "Census Blocks" key, the majority of the TigerWeb data keys are likely to always be absent from the `lookupTable`

https://github.com/vanderbilt-redcap/Census-Tract-Geocoding-External-Module/blob/17520641ca47a27c59759e374c3a952dbd40d9e4/CensusExternalModule.php#L107

e.g. the `CSA` key is only available under `Combined Statistical Areas`; perhaps the API worked differently when this module was created.

A solution to this issue might be to populate `lookupTable` by checking each key for `geographies` in reverse order until a value is found.  
Although this particular solution doesn't hold for values only contained in `addressComponents` (i.e. `zip`). Additionally, it's not a simple problem to report the geographic bound which the data was sourced from as these data may be piped into validated fields.

----

<details>
<summary>JSON response for 2020 data, links to API</summary>

```json
{
  "result": {
    "input": {
      "address": {
        "address": "2201 West End Ave, Nashville, TN 37235"
      },
      "vintage": {
        "isDefault": false,
        "id": "420",
        "vintageName": "Census2020_Current",
        "vintageDescription": "Census2020 Vintage - Current Benchmark"
      },
      "benchmark": {
        "isDefault": true,
        "benchmarkDescription": "Public Address Ranges - Current Benchmark",
        "id": "4",
        "benchmarkName": "Public_AR_Current"
      }
    },
    "addressMatches": [
      {
        "tigerLine": {
          "side": "L",
          "tigerLineId": "647614301"
        },
        "geographies": {
          "2018 State Legislative Districts - Upper": [
            {
              "POP100": 234648,
              "GEOID": "47021",
              "CENTLAT": "+36.1014194",
              "AREAWATER": 22266480,
              "STATE": "47",
              "BASENAME": "21",
              "OID": "212704487270825",
              "LSADC": "LU",
              "SLDU": "021",
              "FUNCSTAT": "N",
              "INTPTLAT": "+36.1010901",
              "NAME": "State Senate District 21",
              "OBJECTID": 177,
              "CENTLON": "-086.6883112",
              "LSY": "2018",
              "AREALAND": 240041777,
              "INTPTLON": "-086.6865778",
              "HU100": 102902,
              "MTFCC": "G5210",
              "LDTYP": "O"
            }
          ],
          "States": [
            {
              "STATENS": "01325873",
              "POP100": 6910840,
              "GEOID": "47",
              "CENTLAT": "+35.8582545",
              "AREAWATER": 2344232327,
              "STATE": "47",
              "BASENAME": "Tennessee",
              "STUSAB": "TN",
              "OID": "2747051118086",
              "LSADC": "00",
              "FUNCSTAT": "A",
              "INTPTLAT": "+35.8584600",
              "DIVISION": "6",
              "NAME": "Tennessee",
              "REGION": "3",
              "OBJECTID": 13,
              "CENTLON": "-086.3490839",
              "AREALAND": 106770495857,
              "INTPTLON": "-086.3496339",
              "HU100": 3031605,
              "MTFCC": "G4000",
              "UR": "M"
            }
          ],
          "Combined Statistical Areas": [
            {
              "POP100": 2118233,
              "GEOID": "400",
              "CENTLAT": "+35.9686285",
              "AREAWATER": 204133347,
              "BASENAME": "Nashville-Davidson--Murfreesboro, TN",
              "OID": "2617013782255192",
              "LSADC": "M0",
              "FUNCSTAT": "S",
              "INTPTLAT": "+35.9686460",
              "NAME": "Nashville-Davidson--Murfreesboro, TN CSA",
              "OBJECTID": 8,
              "CSA": "400",
              "CENTLON": "-086.7044433",
              "INTPTLON": "-086.7069821",
              "AREALAND": 18533025506,
              "HU100": 880654,
              "MTFCC": "G3100"
            }
          ],
          "2020 Urban Areas - Corrected": [
            {
              "GEOID": "61273",
              "CENTLAT": "+36.1264038",
              "AREAWATER": 13731644,
              "BASENAME": "Nashville-Davidson, TN",
              "OID": "27021134432064",
              "UA": "61273",
              "LSADC": "67",
              "FUNCSTAT": "S",
              "INTPTLAT": "+36.1259094",
              "NAME": "Nashville-Davidson, TN Urban Area",
              "OBJECTID": 281,
              "CENTLON": "-086.6924185",
              "AREALAND": 1514962957,
              "INTPTLON": "-086.6911686",
              "MTFCC": "G3500"
            }
          ],
          "County Subdivisions": [
            {
              "COUSUB": "93268",
              "POP100": 17991,
              "GEOID": "4703793268",
              "CENTLAT": "+36.1324166",
              "AREAWATER": 0,
              "STATE": "47",
              "BASENAME": "18",
              "OID": "276702969536411",
              "LSADC": "28",
              "FUNCSTAT": "N",
              "INTPTLAT": "+36.1324166",
              "NAME": "District 18",
              "OBJECTID": 25242,
              "CENTLON": "-086.8039191",
              "COUSUBCC": "Z1",
              "AREALAND": 6351030,
              "INTPTLON": "-086.8039191",
              "HU100": 6227,
              "MTFCC": "G4040",
              "COUSUBNS": "02464407",
              "UR": "U",
              "COUNTY": "037"
            }
          ],
          "Incorporated Places": [
            {
              "DISP_CLR": 4,
              "NECTAPCI": "N",
              "POP100": 689447,
              "GEOID": "4752006",
              "CENTLAT": "+36.1713872",
              "AREAWATER": 55980697,
              "BASENAME": "Nashville-Davidson metropolitan government (balance)",
              "STATE": "47",
              "OID": "27870134387655",
              "LSADC": "00",
              "INTPTLAT": "+36.1718001",
              "PLACE": "52006",
              "FUNCSTAT": "F",
              "NAME": "Nashville-Davidson metropolitan government (balance)",
              "OBJECTID": 12164,
              "PLACECC": "C8",
              "CENTLON": "-086.7841719",
              "CBSAPCI": "Y",
              "AREALAND": 1232288496,
              "HU100": 316363,
              "INTPTLON": "-086.7850016",
              "PLACENS": "02405092",
              "MTFCC": "G4110",
              "UR": "M"
            }
          ],
          "Counties": [
            {
              "POP100": 715884,
              "GEOID": "47037",
              "CENTLAT": "+36.1694525",
              "AREAWATER": 56439483,
              "STATE": "47",
              "BASENAME": "Davidson",
              "OID": "27570134377631",
              "LSADC": "06",
              "FUNCSTAT": "C",
              "INTPTLAT": "+36.1691287",
              "NAME": "Davidson County",
              "OBJECTID": 1809,
              "CENTLON": "-086.7847147",
              "COUNTYCC": "H6",
              "COUNTYNS": "01639737",
              "AREALAND": 1304577371,
              "INTPTLON": "-086.7847898",
              "HU100": 328309,
              "MTFCC": "G4020",
              "UR": "M",
              "COUNTY": "037"
            }
          ],
          "2018 State Legislative Districts - Lower": [
            {
              "POP100": 69925,
              "GEOID": "47055",
              "CENTLAT": "+36.1317908",
              "SLDL": "055",
              "AREAWATER": 555470,
              "STATE": "47",
              "BASENAME": "55",
              "OID": "213704487271044",
              "LSADC": "LL",
              "FUNCSTAT": "N",
              "INTPTLAT": "+36.1315419",
              "NAME": "State House District 55",
              "OBJECTID": 1963,
              "CENTLON": "-086.8633605",
              "LSY": "2018",
              "AREALAND": 58072372,
              "INTPTLON": "-086.8642735",
              "HU100": 34157,
              "MTFCC": "G5220",
              "LDTYP": "O"
            }
          ],
          "116th Congressional Districts": [
            {
              "POP100": 800247,
              "GEOID": "4705",
              "CENTLAT": "+36.1771610",
              "CDSESSN": "116",
              "AREAWATER": 71798552,
              "STATE": "47",
              "BASENAME": "5",
              "OID": "211704487269645",
              "LSADC": "C2",
              "FUNCSTAT": "N",
              "INTPTLAT": "+36.1873306",
              "NAME": "Congressional District 5",
              "OBJECTID": 114,
              "CENTLON": "-087.0699203",
              "AREALAND": 3233151436,
              "INTPTLON": "-087.0740252",
              "HU100": 363163,
              "CD116": "05",
              "MTFCC": "G5200"
            }
          ],
          "Census Tracts": [
            {
              "POP100": 4812,
              "GEOID": "47037016500",
              "CENTLAT": "+36.1456055",
              "AREAWATER": 0,
              "STATE": "47",
              "BASENAME": "165",
              "OID": "20770215796597",
              "LSADC": "CT",
              "FUNCSTAT": "S",
              "INTPTLAT": "+36.1456055",
              "NAME": "Census Tract 165",
              "OBJECTID": 57178,
              "TRACT": "016500",
              "CENTLON": "-086.8050100",
              "AREALAND": 1494439,
              "INTPTLON": "-086.8050100",
              "HU100": 1094,
              "MTFCC": "G5020",
              "UR": "U",
              "COUNTY": "037"
            }
          ],
          "Census Blocks": [
            {
              "GEOID": "470370165002000",
              "STATE": "47",
              "BASENAME": "2000",
              "LSADC": "BK",
              "INTPTLAT": "+36.1463730",
              "OBJECTID": 664024,
              "BLKGRP": "2",
              "AREALAND": 354033,
              "HU100": 3,
              "VINTAGE": "70",
              "LWBLKTYP": "L",
              "UR": "U",
              "COUNTY": "037",
              "SUFFIX": "",
              "TABBLKSUFX2": "",
              "POP100": 1906,
              "CENTLAT": "+36.1463730",
              "BLOCK": "2000",
              "AREAWATER": 0,
              "OID": "210701007549205",
              "FUNCSTAT": "S",
              "NAME": "Block 2000",
              "TRACT": "016500",
              "CENTLON": "-086.8023381",
              "ACT": "",
              "INTPTLON": "-086.8023381",
              "MTFCC": "G5040"
            }
          ]
        },
        "coordinates": {
          "x": -86.80349503726946,
          "y": 36.14929681521341
        },
        "addressComponents": {
          "zip": "37240",
          "streetName": "END",
          "preType": "",
          "city": "NASHVILLE",
          "preDirection": "W",
          "suffixDirection": "",
          "fromAddress": "2201",
          "state": "TN",
          "suffixType": "AVE",
          "toAddress": "2299",
          "suffixQualifier": "",
          "preQualifier": ""
        },
        "matchedAddress": "2201 W END AVE, NASHVILLE, TN, 37240"
      }
    ]
  }
}
```

Data examples:
2020  
https://geocoding.geo.census.gov/geocoder/geographies/onelineaddress?address=2201%20West%20End%20Ave%2C%20Nashville%2C%20TN%2037235&benchmark=Public_AR_Current&vintage=Census2020_Current&format=json

2010  
https://geocoding.geo.census.gov/geocoder/geographies/onelineaddress?address=2201%20West%20End%20Ave%2C%20Nashville%2C%20TN%2037235&benchmark=Public_AR_Current&vintage=Census2010_Current&format=json
</details>

### PR DESCRIPTION
**Pull request by ChemiKyle**
This constitutes a major version change (although it does still work with the previous version).

Due to recent breakages with API parameters, rather than selecting a census year, users must select a benchmark and vintage from which to receive data. The module will now dynamically fetch currently available benchmark/vintage pairs from the US Census API.

Closes #3  
Due to the `Census Blocks` geographic bound not appearing in some vintages (some of them prepend the year), desired fields are searched for in sequentially increasing geographic areas (default ordering is by land area, based on the - possibly dubious - assumption that people live on land).

QoL:
- Moved all the javascript into its own file
- Bumped framework version up to 16

---

I will make an additional commit updating the README, I'll also provide an example XML and module settings.  
That will have to wait until Monday because my brain is mush, I used the last of my weekly allotment of linguistic ability to write this excuse.

### CODE REVIEW

File: CensusExternalModule.php [hunk 1]
```diff
@@ -94,117 +147,31 @@ function getSharedArgs($censusYear){
...
 
 	function addScript($project_id, $record, $instrument, $event_id, $group_id, $survey_hash = null, $response_id = null) {
-		if ($project_id) {
-			$addressField = $this->getProjectSetting('address');
-			$latitudeField = $this->getProjectSetting('latitude');
-			$longitudeField = $this->getProjectSetting('longitude');
+		if (!$project_id) { return; }
+		$this->initializeJavascriptModuleObject();
+
+		echo "<script> const module = " . $this->getJavascriptModuleObjectName() . "</script>";
```
Comment (mmcev106) [inline CensusExternalModule.php:153]: Defining anything in JS top level scope is discouraged in module development.  This one I think is especially concerning as `module` is very generic and can easily conflict with other modules.  There are a few different ways to do this safely.  If you'd like to discuss those, or if it isn't clear why this is something we should always consider during development, please give me call.

**Review by mmcev106 — CHANGES_REQUESTED** [review]
@ChemiKyle, this looks great overall.  Apart from my comment, the only other concern is around you mentioning that updating the major version would be appropriate, even though old versions would still work.  Would you be able to clarify?  My main concern is to avoid project configuration changes if at all possible, as that will end up being more work for you.

File: CensusExternalModule.php (hunk 1 above)
Comment (ChemiKyle) [reply CensusExternalModule.php:153]: Good catch, that wasn't meant to make it into the commit! No discussion needed, I'm well aware of the collision potential there.  
Quick sanity check, setting `const module = <JSMOName>` in the anonymous function used in `$(document).ready( () => {} )` is safely confined within the scope of that function, right?  
https://github.com/vanderbilt-redcap/Census-Tract-Geocoding-External-Module/pull/7/files#diff-431e5adb210d9982cdf7bacdacad1bab1cd70c57028cb41316d8bb79d2d8b046R4

**Comment by ChemiKyle** [comment]
Regarding major version change: my reasoning is that I'm deprecating an option in the configuration settings, replacing it with a new option. While I've added logic to process extant values for the deprecated option, the new pathway is preferred in the logic and I don't intend to maintain the logic for the deprecated option.  
Since the old option is setting parameters for an API which has regular flux, the processing of those options will _eventually_ break.

**Comment by mmcev106** [comment]
What is the transition plan for projects that use this module?

**Comment by ChemiKyle** [comment]
The easiest for them would be for me to add logic to transition them to a benchmark/vintage that equates to their `year` option. I considered doing this pre-emptively with an adjustment to the `transitionOldSettings` function but left it out since I didn't want to disrupt my ability to test the old path. I can add that back, now that I'm not entrenched in the problem I realize the solution is rather trivial.  
The easiest path for me is that they notice when failures begin to occur and select a benchmark/vintage that meets their needs.

**Comment by mmcev106** [comment]
@ChemiKyle, I believe projects that use this are configured with hard census data that won't change (2010 or 2020).  Can you confirm that those projects will continue functioning as expected with no changes needed?

**Comment by ChemiKyle** [comment]
Yes, if the new `benchmark_vintage` configuration is not set, the `year` will be processed as usual, that's handled by these lines:

https://github.com/vanderbilt-redcap/Census-Tract-Geocoding-External-Module/pull/7/files#diff-1cff3038c55e93ba1f53e6ac162b35d0d148b0f86eb260b1fabc9bf8118f3934R9-R13

https://github.com/vanderbilt-redcap/Census-Tract-Geocoding-External-Module/pull/7/files#diff-0ccc8949cebe86f8816a2a07dd72559334b4718cc7883f8bbf12a0d89240c200R8-R12

**Review by mmcev106 — APPROVED** [review]
That all sounds good to me.  Approved!

**Pull request merged**